src/content.js

```diff
@@ -1,7 +1,79 @@
-document.addEventListener('DOMContentLoaded', function() {
-  var elOpenSearch = document.querySelector('[type="application/opensearchdescription+xml"]');
-  if(elOpenSearch) {
-    elOpenSearch.remove();
-    chrome.runtime.sendMessage("opensearch-block");
-  }  
-}, false);
\ No newline at end of file
+
+// Add an extra child input to any form that only has one
+function spoilFormGet(elem) {
+ console.info({Found: elem});
+
+ // Bail early if it's already one Chrome won't autodetect
+ if( (!/^http/i.test(elem.getAttribute('action'))) &&
+     (!/^http/i.test(elem.action))
+ ) {
+  return;
+ }
+  // Need to check this here rather than in the selector since elem.action
+  // is a full URL (in my testing) even if the form specifies
+  // action="/whatever/path".
+  // Use getAttribute() since otherwise <input name="action"> is exposed as
+  // elem.action.
+
+ if( (String(elem.getAttribute('method')).toLowerCase() !== 'get') &&
+     (String(elem.method).toLowerCase() !== 'get')
+ ) {
+  return;
+ }
+  // Ditto - have to check here in case the form doesn't expressly specify
+  // a method
+
+ var texts = elem.querySelectorAll(':scope input[type="text" i]');
+ var searches = elem.querySelectorAll(':scope input[type="search" i]');
+ var onetext = (texts.length === 1 && searches.length === 0);
+ var onesearch = (texts.length === 0 && searches.length === 1);
+ if( !(onetext || onesearch) ) return;
+
+ if(elem.querySelector(':scope input[type="password" i]')) return;
+ if(elem.querySelector(':scope input[type="file" i]')) return;
+ if(elem.querySelector(':scope textarea')) return;
+
+ // Add a <textarea> - unlike <input>, it doesn't block implicit submission
+ // per https://www.tjvantoll.com/2013/01/01/enter-should-submit-forms-stop-messing-with-that/
+ var newelem;
+ newelem = document.createElement('textarea');
+ newelem.name = 'chrome_dont_add_custom_search_engines_srsly';
```

Comment:
Honestly?   
Replacing an annoying behavior by another one is *not* an improvement...

Reply:
hey @zessx , I agree that adding a textarea is a bit of an "aggressive" method, although it does help in stopping new search engines from being added

do you feel like this approach might "break" many web forms? or is it more a consideration of "do as little harm to the DOM as possible"?

I'm wondering if we couple separate the form detection/textarea adding and mark it as an "advanced" feature which could be disabled by default

thoughts? @cxw42 ?

Reply:
The [autodetection](https://www.chromium.org/tab-to-search), part 2, is actually very effective at automatically adding most search boxes :/ .  No surprise, since that's what it's supposed to do! :)  Take out the `<textarea>`, and you're right back to where the extension was before.

My first try was adding an `onsubmit` script, since the autodetection page mentions onsubmit scripts aren't autodetected.  However, that didn't work.  I am guessing that so many forms now have onsubmit that the Chromium team removed that check, and just didn't update the documentation page.

My second try was to add a hidden `<input>`, but that prevents GitHub's search forms from working (pressing Enter no longer submits when there are multiple `<input>` elements).

@zessx If you can show me a better way that also works, please do!  I have started watching this repo and will be looking forward to your PR ;) .

Reply:
I was more upset by the 12-miles-long GET parameter name than by the addition of a `textarea`.   

We may not have a better way right know, and I'm fine with it, but at least let's choose a shorter name. I work a lot with URLs (on search engines, specifically), and to get `&chrome_dont_add_custom_search_engines_srsly=` in all of them is a big no.   
What about `cdacse`, or `dacse`?

On the `textarea` itself, I think it *may* interferes with abstract APIs (taking into account every sent parameters). Even if this interference is a proof that this kind of APIs are highly vulnerable, it could be a good idea to get a whitelist in the extension.

Reply:
I agree that the long GET URLs are "visually" annoying, and can get in the way of testing/development.

I think that one good side about using a unique name for that parameter is that it is highly google-able, and won't risk conflicting with internal parameters which may use names like `cdacse`.

Having a whitelist seems like a very good idea. Something akin to ad blockers where clicking the extension icon would allow you turning it on/off for the domain, and greying out the icon when it detects that it's on a whitelisted site..?

Reply:
I think we can find something almost unique, *and* shorter than 43 chars :)

What about a combination of a little name, and a time-related number?
This has a really low level of collision, and in the worst case it would work again later.

Time-related part | Validity | Collision level | Exemple
---|---|---|---
minutes | 1 minute | medium | `cse09` 
hour+minutes | 1 minute | low | `cse1509` 
6-digit timestamp | 3 hours~ | low | `cse152173`

Reply:
Does it have to be sent in order to foil Chrome’s detection? ’Cause you can prevent it from being sent by setting the `disabled` property to true.

Reply:
good idea, @ap -- I just tried disabling the textarea and unfortunately that lets the search engine be added.

Reply:
@zessx Thanks for clarifying what you were referring to in your first message.  I agree with the comment out-of-line from @gregsadetsky setting whitelist as a higher priority than name-change.

I haven't been a server admin for a long time, but I would think being able to easily identify something weird in my logs would make me happy.  And the existing name is already indexed on Google :) .

Reply:
@gregsadetsky You beat me to it :) .  I also just tried the same - bleepingcomputer is added with `<textarea disabled>`.

Reply:
Hmm. Then maybe like this?

````html
<input type="radio" name="chrome_dont_add_custom_search_engines_srsly" tabindex="-1" style="display: none !important">
````

The trick to this is that a `radio` (or `checkbox`) without a `checked` attribute doesn’t become part of the URL… so this has the same effect as marking it `disabled` – but hopefully while still foiling Chrome’s detection.

The other attributes try to make it as unobtrusive as possible. Submitting the form using Enter still works for me if I use the DOM inspector to add such an `<input>` to the GitHub search form at the top of the page, at least.

Reply:
@ap Good idea.  Unfortunately, my [test branch](https://github.com/cxw42/chrome-dont-add-custom-search-engines/tree/radio) fails on bleepingcomputer and on gcmap.com.

Reply:
digging around, I found [ParseIfNecessary](https://github.com/chromium/chromium/blob/master/components/search_engines/template_url.cc#L811) which counts the "search terms" found in the URL and sets the bool `supports_replacements_` which determines if the URL is then used to create a new custom search engine or not.

this would confirm that we do need to send more than 1 value to foil the search engine detection

Reply:
Good find.  Wild thought - what about an astral or private-use field name?  If we assume sites are 8-bit clean unless whitelisted, we could use U+10cda U+1005e or something not yet assigned.  In the query-string, that will work out as escaped utf8 bytes.

Reply:
I'd err on the side of caution / sites not necessarily being ready for this kind of input? + the current google-ability of the name used for the textarea seems like a good idea

love the idea, just unsure of repercussions...!